File: Server_Root/Server/RoomServer/Entity/Entities/Character/Character.cpp
#include "stdafx.h"
#include "Character.h"
#include "../../../MasterData/MasterDataManager.h"
#include "../../../MasterData/Behavior.h"
#include "../../../MasterData/Character.h"
#include "../../../Factory.h"
#include "../Projectile/Projectile.h"
#include "../../../Game/BaeGameRoom.h"
#include "../../../Behavior/BehaviorIDs.h"
#include "../../../Behavior/Behaviors/Die.h"
#include "../../../State/StateIDs.h"
#include "../../../GameEvent/GameEvents/CharacterAttack.h"
#include "../../../GameEvent/GameEvents/CharacterRespawn.h"

Character::Character(BaeGameRoom* pGameRoom, int nID, int nMasterDataID, Role role) : IEntity(pGameRoom, nID, nMasterDataID)
{
	m_Role = role;
}

Character::~Character()
{
	for (list<ISkill*>::iterator it = m_listSkill.begin(); it != m_listSkill.end(); ++it)
	{
		delete *it;
	}
	m_listSkill.clear();
}

void Character::SetRole(Role role)
{
	m_Role = role;
}

Character::Role Character::GetRole()
{
	return m_Role;
}

void Character::Initialize()
{
	MasterData::Character* pMasterCharacter = NULL;
	MasterDataManager::Instance()->GetData<MasterData::Character>(m_nMasterDataID, pMasterCharacter);

	m_fSize = pMasterCharacter->m_fSize;
	m_fHeight = pMasterCharacter->m_fHeight;
	m_fDefault_Y = pMasterCharacter->m_fDefault_Y;
	m_nDefaultBehaviorID = pMasterCharacter->m_nDefaultBehaviorID;

	for (vector<int>::iterator it = pMasterCharacter->m_vecSkillID.begin(); it != pMasterCharacter->m_vecSkillID.end(); ++it)
	{
		ISkill* pSkill = Factory::Instance()->CreateSkill(m_pGameRoom, this, *it);
		if (pSkill != NULL)
		{
			pSkill->Initialize();
			m_listSkill.push_back(pSkill);
		}
	}

	for (vector<int>::iterator it = pMasterCharacter->m_vecBehaviorID.begin(); it != pMasterCharacter->m_vecBehaviorID.end(); ++it)
	{
		IBehavior* pBehavior = Factory::Instance()->CreateBehavior(m_pGameRoom, this, *it);
		if (pBehavior != NULL)
		{
			pBehavior->Initialize();
			m_listBehavior.push_back(pBehavior);
		}
	}
}

float Character::GetSpeed()
{
	return m_CurrentStatus.m_fSpeed * (m_fSpeedPercent / 100);
}

float Character::GetMaximumSpeed()
{
	return m_CurrentStatus.m_fMaximumSpeed;
}

FBS::Data::EntityType Character::GetEntityType()
{
	return FBS::Data::EntityType::EntityType_Character;
}

void Character::NotifyGameEvent(IGameEvent* pGameEvent)
{
}

bool Character::IsTerrainPassable()
{
	return false;
}

int Character::GetMoveCollisionTypes()
{
	int nTypes = CollisionObject::Type::CollisionObjectType_None;

	if (m_Role == Character::Role::Challenger)
	{
		nTypes = CollisionObject::Type::CollisionObjectType_Terrain | CollisionObject::Type::CollisionObjectType_Character_Disturber | CollisionObject::Type::CollisionObjectType_Projectile
			| CollisionObject::Type::CollisionObjectType_Item;
	}
	else if (m_Role == Character::Role::Disturber)
	{
		nTypes = CollisionObject::Type::CollisionObjectType_Terrain| CollisionObject::Type::CollisionObjectType_Character_Challenger;
	}

	return nTypes;
}

void Character::UpdateSkills(long long lUpdateTime)
{
	list<ISkill*> listSkill = GetActivatedSkills();
	for (list<ISkill*>::iterator it = listSkill.begin(); it != listSkill.end(); ++it)
	{
		ISkill* pSkill = *it;
		if (pSkill != NULL)
			pSkill->Update(lUpdateTime);

		if (m_bDestroyReserved)
			break;
	}
}

void Character::LateUpdate(long long lUpdateTime)
{
	TrimState();

	if (m_nDefaultBehaviorID != -1 && !IsBehavioring() && GetBehavior(m_nDefaultBehaviorID) != NULL && IsAlive())
	{
		GetBehavior(m_nDefaultBehaviorID)->Start(lUpdateTime);
	}
}

list<ISkill*> Character::GetAllSkills()
{
	return m_listSkill;
}

list<ISkill*> Character::GetActivatedSkills()
{
	list<ISkill*> listActivatedSkill;

	for (list<ISkill*>::iterator it = m_listSkill.begin(); it != m_listSkill.end(); ++it)
	{
		if ((*it)->IsActivated())
			listActivatedSkill.push_back(*it);
	}

	return listActivatedSkill;
}

bool Character::IsSkilling()
{
	for (list<ISkill*>::iterator it = m_listSkill.begin(); it != m_listSkill.end(); ++it)
	{
		if ((*it)->IsActivated())
			return true;
	}

	return false;
}

void Character::InitStatus(CharacterStatus status)
{
	m_OriginalStatus = status;
	m_CurrentStatus = status;
}

float Character::GetCurrentMP()
{
	return m_CurrentStatus.m_nMP;
}

void Character::SetCurrentMP(float fMP)
{
	m_CurrentStatus.m_nMP = fMP;
}

void Character::SetMoveSpeed(float fSpeed)
{
	m_CurrentStatus.m_fSpeed = fSpeed;
}

void Character::PlusMoveSpeed(float fValue)
{
	m_CurrentStatus.m_fSpeed += fValue;
}
void Character::MinusMoveSpeed(float fValue)
{
	m_CurrentStatus.m_fSpeed -= fValue;
}

bool Character::IsAlive()
{
	return m_CurrentStatus.m_nHP > 0;
}

void Character::OnMoved(int nDistance)
{
	int nBonusPoint = 100;

	m_CurrentStatus.m_fMovePoint += (nDistance * m_CurrentStatus.m_fMPChargeRate);

	if (m_CurrentStatus.m_fMovePoint >= nBonusPoint)
	{
		int nGetCount = m_CurrentStatus.m_fMovePoint / nBonusPoint;

		m_CurrentStatus.m_fMovePoint -= (nGetCount * nBonusPoint);
		m_CurrentStatus.m_nMP += nGetCount;
	}

	//	GameEvent status change
}

void Character::OnAttacked(int nAttackingEntityID, int nDamage, long long lTime)
{
	if (HasCoreState(CoreState::CoreState_Invincible) || !IsAlive())
		return;

	m_CurrentStatus.m_nHP -= nDamage;

	GameEvent::CharacterAttack* pCharacterAttack = new GameEvent::CharacterAttack();
	pCharacterAttack->m_fEventTime = lTime / 1000.0f;
	pCharacterAttack->m_nAttackingEntityID = nAttackingEntityID;
	pCharacterAttack->m_nAttackedEntityID = m_nID;
	pCharacterAttack->m_nDamage = nDamage;

	m_pGameRoom->AddGameEvent(pCharacterAttack);

	if (m_CurrentStatus.m_nHP <= 0)
	{
		IBehavior* pDieBehavior = GetBehavior(BehaviorID::DIE);

		pDieBehavior->Start(lTime);
		pDieBehavior->Update(lTime);
	}
}

void Character::OnRespawn(long long lTime)	//	last position?
{
	m_CurrentStatus.m_nHP = m_OriginalStatus.m_nHP;

	GameEvent::CharacterRespawn* pCharacterRespawn = new GameEvent::CharacterRespawn();
	pCharacterRespawn->m_fEventTime = lTime / 1000.0f;
	pCharacterRespawn->m_nEntityID = m_nID;
	pCharacterRespawn->m_vec3Position = GetPosition();

	m_pGameRoom->AddGameEvent(pCharacterRespawn);

	IState* pState = Factory::Instance()->CreateState(m_pGameRoom, this, StateID::RespawnInvincible, lTime);
	pState->Initialize();
	AddState(pState, lTime);
	pState->Update(lTime);
}

File: Client_Root/Client/Assets/Scripts/Entity/Entities/Character/Character.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Character : IEntity
{
    public enum Role
    {
        Challenger = 0,
        Disturber,
    };

    private Role m_Role = Role.Challenger;

    //    protected List<ISkill> m_listSkill;
    private Dictionary<int,int> m_dicGameItemEffect = new Dictionary<int,int>();
    private GameItem[] m_GameItems = new GameItem[2];

    protected CharacterStatus m_OriginalStatus;
    protected CharacterStatus m_CurrentStatus;

    private int m_nJumpRegenerationTickInterval = 0;
    private int m_nJumpRegenerationTick = 0;

    public float m_fSpeedPercent = 100;

    private float m_fBestHeight = 0;

    public void SetRole(Role role)
    {
        m_Role = role;
    }

    public Role GetRole()
    {
        return m_Role;
    }

    public override void Initialize(int nID, int nMasterDataID, params object[] param)
    {
        m_nID = nID;
        m_nMasterDataID = nMasterDataID;
        m_Role = (Role)param[0];
		InitStatus((CharacterStatus)param[1]);

		m_nJumpRegenerationTickInterval = (int)(m_CurrentStatus.m_fJumpRegenerationTime / BaeGameRoom2.Instance.GetTickInterval());

        MasterData.Character masterCharacter = null;
        MasterDataManager.Instance.GetData<MasterData.Character>(m_nMasterDataID, ref masterCharacter);

        //        foreach(int nSkillID in pMasterCharacter.m_listSkillID)
        //        {
        //            Factory.Instance.CreateBehavior(this, nSkillID);
        //        }

        foreach(KeyValuePair<int, int> kv in masterCharacter.m_dicGameItemEffect)
        {
            m_dicGameItemEffect.Add(kv.Key, kv.Value);
        }

        foreach(int nBehaviorID in masterCharacter.m_listBehaviorID)
        {
            IBehavior behavior = Factory.Instance.CreateBehavior(nBehaviorID);

            if (behavior == null)
            {
                Debug.LogWarning("behavior is null! nBehaviorID : " + nBehaviorID);
                continue;
            }

            behavior.Initialize(this, nBehaviorID, BaeGameRoom2.Instance.GetTickInterval());

            m_listBehavior.Add(behavior);
        }

        GameObject goEntityUI = ObjectPool.Instance.GetGameObject("CharacterModel/EntityUI");
        goEntityUI.transform.parent = gameObject.transform;

        EntityUI entityUI = goEntityUI.GetComponent<EntityUI>();
        entityUI.Initialize(FBS.Data.EntityType.Character, nID, nMasterDataID);

        m_EntityUI = entityUI;
    }

    public override float GetSpeed()
    {
        return m_CurrentStatus.m_fSpeed * (m_fSpeedPercent / 100);
    }

    public override float GetMaximumSpeed()
    {
        return m_CurrentStatus.m_fMaximumSpeed;
    }

    public override FBS.Data.EntityType GetEntityType()
    {
        return FBS.Data.EntityType.Character;
    }

    public override void NotifyGameEvent(IGameEvent gameEvent)
    {
    }

    public void UpdateSkills(long lUpdateTime)
    {
    }

    public override void LateUpdateWorld(int nUpdateTick)
    {
        m_fBestHeight = Mathf.Max(m_fBestHeight, GetCurrentHeight());   //  Timing correct?

		if (m_CurrentStatus.m_nJumpCount < m_CurrentStatus.m_nMaximumJumpCount)
        {
			m_nJumpRegenerationTick++;

			if (m_nJumpRegenerationTick == m_nJumpRegenerationTickInterval)
            {
				m_CurrentStatus.m_nJumpCount++;
				m_nJumpRegenerationTick = 0;
            }
        }
        else
        {
			m_nJumpRegenerationTick = 0;
        }

        if (HasCoreState(CoreState.CoreState_Faint))
            return;

        if (m_nDefaultBehaviorID != -1 && !IsBehavioring() && GetBehavior(m_nDefaultBehaviorID) != null && IsAlive())
        {
            GetBehavior(m_nDefaultBehaviorID).StartTick(nUpdateTick);
        }
    }

    //    public ISkill* GetSkill(int nID);
    //    public List<ISkill*> GetAllSkills();
    //    public List<ISkill*> GetActivatedSkills();
    //    public bool IsSkilling();

	private void InitStatus(CharacterStatus status)
    {
        m_OriginalStatus = status;
        m_CurrentStatus = status;
    }

    public int GetCurrentMP()
    {
        return m_CurrentStatus.m_nMP;
    }

    public void SetCurrentMP(int nMP)
    {
        m_CurrentStatus.m_nMP = nMP;
    }

    public void SetMoveSpeed(float fSpeed)
    {
        m_CurrentStatus.m_fSpeed = fSpeed;
    }

    public void PlusMoveSpeed(float fValue)
    {
        m_CurrentStatus.m_fSpeed += fValue;
    }

    public void MinusMoveSpeed(float fValue)
    {
        m_CurrentStatus.m_fSpeed -= fValue;
    }

	public int GetJumpCount()
    {
        return m_CurrentStatus.m_nJumpCount;
    }

	public int GetMaximumJumpCount()
    {
        return m_CurrentStatus.m_nMaximumJumpCount;
    }

    public void OnJump(int nTick)
    {
		if (!IsJumpable())
            return;

		m_CurrentStatus.m_nJumpCount--;

        GetBehavior(MasterDataDefine.BehaviorID.JUMP).StartTick(nTick);
    }

    public void OnAttacked(int nAttackingEntityID, int nDamage, int nTick)
    {
        if (HasCoreState(CoreState.CoreState_Invincible) || !IsAlive())
            return;

        m_CurrentStatus.m_nHP -= nDamage;

        if (m_CurrentStatus.m_nHP <= 0)
        {
            List<IBehavior> listActivatedBehavior = GetActivatedBehaviors();
            foreach(IBehavior behavior in listActivatedBehavior)
            {
                behavior.Stop();
            }

            //  Die behavior
            IBehavior dieBehavior = GetBehavior(MasterDataDefine.BehaviorID.DIE);
            dieBehavior.StartTick(nTick);
            dieBehavior.UpdateTick(nTick);

            //  Faint state
            IState state = Factory.Instance.CreateState(MasterDataDefine.StateID.FAINT);
            state.Initialize(this, MasterDataDefine.StateID.FAINT, BaeGameRoom2.Instance.GetTickInterval());

            AddState(state, nTick);

            state.StartTick(nTick);
            state.UpdateTick(nTick);

            BaeGameRoom2.Instance.OnPlayerDie(m_nID, nAttackingEntityID);
        }
    }

    public void OnRespawn(int nTick)
    {
        m_CurrentStatus.m_nHP = m_OriginalStatus.m_nHP;

        IState state = Factory.Instance.CreateState(MasterDataDefine.StateID.RESPAWN_INVINCIBLE);
        state.Initialize(this, MasterDataDefine.StateID.RESPAWN_INVINCIBLE, BaeGameRoom2.Instance.GetTickInterval());

        AddState(state, nTick);

        state.StartTick(nTick);
        state.UpdateTick(nTick);

        BaeGameRoom2.Instance.OnPlayerRespawn(m_nID);
    }

    public void OnMoved(float fDistance, long lTime)
    {
        int nBonusPoint = 100;
        float fMovePoint = fDistance * m_CurrentStatus.m_fMPChargeRate * 100;
        m_CurrentStatus.m_fMovePoint += fMovePoint;
        BaeGameRoom2.Instance.AddCharacterStatusChangeGameEvent(lTime / 1000.0f, m_nID, "MovePoint", "Move", fMovePoint);

        if (m_CurrentStatus.m_fMovePoint >= nBonusPoint)
        {
            int nGetCount = (int)(m_CurrentStatus.m_fMovePoint / (float)nBonusPoint);

            m_CurrentStatus.m_fMovePoint -= (nGetCount * nBonusPoint);
            BaeGameRoom2.Instance.AddCharacterStatusChangeGameEvent(lTime / 1000.0f, m_nID, "MovePoint", "Transfer", -nGetCount * nBonusPoint);

            m_CurrentStatus.m_nMP += nGetCount;
            BaeGameRoom2.Instance.AddCharacterStatusChangeGameEvent(lTime / 1000.0f, m_nID, "MP", "MovePoint", nGetCount);
        }
    }

    public bool IsAlive()
    {
        return m_CurrentStatus.m_nHP > 0;
    }

    public bool IsJumpable()
    {
		return IsAlive() && !HasCoreState(CoreState.CoreState_Faint) &&  m_CurrentStatus.m_nJumpCount > 0;
    }

    public float GetCurrentHeight()
    {
        return m_EntityUI.GetPosition().y;
    }

    public float GetBestHeight()
    {
        return m_fBestHeight;
    }

    public void OnGetGameItem(GameItem item)
    {
        for (int i = 0; i < m_GameItems.Length; ++i)
        {
            if (m_GameItems[i] == null)
            {
                m_GameItems[i] = item;

                if (BaeGameRoom2.Instance.GetUserEntityID() == m_nID)
                {
                    BaeGameRoom2.Instance.OnUserGameItemChanged(m_GameItems);
                }

                return;
            }
        }
    }

    public void OnUseGameItem(int nID)
    {
        GameItem target = null;
        for (int i = 0; i < m_GameItems.Length; ++i)
        {
            if (m_GameItems[i] != null && m_GameItems[i].GetID() == nID)
            {
                target = m_GameItems[i];
                m_GameItems[i] = null;
                break;
            }
        }

        if (target == null)
        {
            Debug.LogError("[OnUseGameItem] target is null! nID : " + nID);
            return;
        }

        if (BaeGameRoom2.Instance.GetUserEntityID() == m_nID)
        {
            BaeGameRoom2.Instance.OnUserGameItemChanged(m_GameItems);
        }

        int nGameItemEffectID = m_dicGameItemEffect[target.GetMasterDataID()];

        MasterData.GameItemEffect masterGameItemEffect = null;
        MasterDataManager.Instance.GetData<MasterData.GameItemEffect>(nGameItemEffectID, ref masterGameItemEffect);

        if (masterGameItemEffect.m_Type == MasterData.GameItemEffect.Type.Behavior)
        {
            IBehavior targetBehavior = GetBehavior(masterGameItemEffect.m_nTargetID);
            targetBehavior.StartTick(BaeGameRoom2.Instance.GetCurrentTick());
            targetBehavior.UpdateTick(BaeGameRoom2.Instance.GetCurrentTick());
        }
    }
}